bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/IRepository.java

```diff
@@ -222,4 +222,13 @@ public interface IRepository<T> extends IAdaptable, IQueryable<T> {
 	 * @since 2.1
 	 */
 	String setProperty(String key, String value, IProgressMonitor monitor);
+
+	/**
+	 * Returns true if this repository contains the given element.
+	 *
+	 * @param element the element to query
+	 * @return true if the given element is already in this repository
+	 * @since 2.8
+	 */
+	boolean contains(T element);
```

Comment:
could this be default implemented?

Reply:
An "everything" query can be like this
```
new ExpressionMatchQuery<>(IInstallableUnit.class, ExpressionUtil.TRUE_EXPRESSION)
```
so I think it would be tricky to create a correct query even for that case though I suppose with some instanceof tests one could do that...

Reply:
what about

`query(QueryUtil.createMatchQuery(ExpressionUtil.TRUE_EXPRESSION)).toSet().contains(element)`

Reply:
See org.eclipse.equinox.p2.query.QueryUtil.ALL_UNITS and note the type of that query:
```
IQuery<IInstallableUnit> ALL_UNITS = QueryUtil.createMatchQuery(ExpressionUtil.TRUE_EXPRESSION);
```

Reply:
The problem is that here it is operated on the generic type T and not `IInstallableUnit` and I didn't found a way to create a query for any type (generic type probably won't work since that info is lost at runtime).

But it could be default implemented in the IMetadataRepository.

Reply:
As far as I can see the class is not actually used if expression is true, but maybe its even simpler:

```
return !query.query(new IQuery<T>() {

			@Override
			public IQueryResult<T> perform(Iterator<T> iterator) {

				while(iterator.hasNext()) {
					T t = iterator.next();
					if(Objects.equals(t, item)) {
						return new CollectionResult<>(List.of(t));
					}
				}
				return new CollectionResult<>(List.of());
			}

			@Override
			public IExpression getExpression() {

				return ExpressionUtil.TRUE_EXPRESSION;
			}
		}, null).isEmpty();
```

Reply:
Thinking a bit more I now think it would even be better to have  	`boolean contains(T element); ` in the `IQueryable` interface with the last provided default implementation...

Reply:
I also tought about moving contains() to queryable but discarded it due to the high implementation work.

But with a good default impl this would be much less work and it can be overriden where suitable.
I'll try it this evening.

Reply:
Added it now and also provided more direct implementations in most complementing classes.

What do you think about keeping the contains(boolean) override in the `IRepository` in order to provide a bit better javadoc?
Unfortunately we cannot remove the contains method from the `IArtifactRepository` interface since this would be a binary incompatible change.

Reply:
Looks good to me if it helps to have better javadoc why not, but I think you need to default implement (and delegate to IQueryable impl) this as well as otherwhise you remove the default impl inherited from the parent otherwhise.

Reply:
It also should have an `@Override` if it's already declared in the base interface.  And it would seem odd to add a method to API without an `@since`; I don't think I've ever seen an `@since` that includes a micro version.  I mention this because of the other comment about there being no additional API; perhaps/probably a default method would additional API.

Reply:
> Looks good to me if it helps to have better javadoc why not, but I think you need to default implement (and delegate to IQueryable impl) this as well as otherwhise you remove the default impl inherited from the parent otherwhise.

Good point. Done that.



> It also should have an `@Override` if it's already declared in the base interface.

Yes that, that's right as well. Done, too.

> And it would seem odd to add a method to API without an `@since`; I don't think I've ever seen an `@since` that includes a micro version. I mention this because of the other comment about there being no additional API; perhaps/probably a default method would additional API.

Agree on that, lets see if the API tools are now fine with it.